src/openjd/adaptor_runtime/_background/frontend_runner.py

```diff
@@ -338,51 +414,84 @@ def _sigint_handler(self, signum: int, frame: Optional[FrameType]) -> None:
         self.cancel()
 
 
-def _load_connection_settings(path: str) -> ConnectionSettings:
-    try:
-        with open(path) as conn_file:
-            loaded_settings = json.load(conn_file)
-    except OSError as e:
-        _logger.error(f"Failed to open connection file: {e}")
-        raise
-    except json.JSONDecodeError as e:
-        _logger.error(f"Failed to decode connection file: {e}")
-        raise
-    return DataclassMapper(ConnectionSettings).map(loaded_settings)
-
-
-def _wait_for_file(filepath: str, timeout_s: float, interval_s: float = 1) -> None:
+def _wait_for_connection_file(
+    filepath: str, max_retries: int, interval_s: float = 1
+) -> ConnectionSettings:
     """
-    Waits for a file at the specified path to exist and to be openable.
+    Waits for a connection file at the specified path to exist, be openable, and have connection settings.
 
     Args:
         filepath (str): The file path to check.
-        timeout_s (float): The max duration to wait before timing out, in seconds.
+        max_retries (int): The max number of retries before timing out.
         interval_s (float, optional): The interval between checks, in seconds. Default is 0.01s.
 
     Raises:
         TimeoutError: Raised when the file does not exist after timeout_s seconds.
     """
+    wait_for(
+        description=f"File '{filepath}' to exist",
+        predicate=lambda: os.path.exists(filepath),
+        interval_s=interval_s,
+        max_retries=max_retries,
+    )
 
-    def _wait():
-        if time.time() - start < timeout_s:
-            time.sleep(interval_s)
-        else:
-            raise TimeoutError(f"Timed out after {timeout_s}s waiting for file at {filepath}")
-
-    start = time.time()
-    while not os.path.exists(filepath):
-        _wait()
+    # Wait before opening to give the backend time to open it first
+    time.sleep(interval_s)
 
-    while True:
-        # Wait before opening to give the backend time to open it first
-        _wait()
+    def file_is_openable() -> bool:
         try:
             open(filepath, mode="r").close()
-            break
         except IOError:
             # File is not available yet
-            pass
+            return False
+        else:
+            return True
+
+    wait_for(
+        description=f"File '{filepath}' to be openable",
+        predicate=file_is_openable,
+        interval_s=interval_s,
+        max_retries=max_retries,
+    )
+
+    def connection_file_loadable() -> bool:
+        try:
+            ConnectionSettingsFileLoader(Path(filepath)).load()
+        except Exception:
+            return False
+        else:
+            return True
+
+    wait_for(
```

Comment:
Hah, nice. I actually just filed a bug for this. It looked like we've seen some data races on windows where the file can be opened, but hadn't yet contained data. Thanks for fixing!